Add overwrite option to propagate state

eslickj opened the pull request:
## Fixes issue #522 

This adds an options to propagate state to overwrite fixed values.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 reviewed (approved):
Looks good to me.

dallan-keylogic commented on idaes/core/util/initialization.py:
```diff
@@ -189,7 +191,7 @@ def propagate_state(destination=None, source=None, arc=None,
         try:
             for i in v:
-                if not (v[i].is_variable_type() and v[i].fixed):
+                if v[i].is_variable_type() and ((not v[i].fixed) or overwrite_fixed):
```
The behavior got changed for when `v[i].is_variable_type()` is false. Are there things for which `v[i].is_variable_type()` evaluates to false for which we still want to set a value @andrewlee94  @bknueven ?

bknueven replied:
The error message below speaks about mutable Params, which have a `value` attribute but no `fixed` attribute. To maintain the existing behavior when `overwrite_fixed` is `False` we should do something like
```suggestion
                if overwrite_fixed or not(v[i].is_variable_type() and v[i].fixed):
```

dallan-keylogic replied:
Very good. Hopefully @eslickj will add a test to ensure this behavior is maintained.

andrewlee94 replied:
From an IDAES specific stand-point, I have to ask why a `Param` would appear in a `Port` - this would tend to indicate a (very big) modeling error on the part of the user. That said, that is something that should be caught well before this point, which would mean when the `Port` is constructed.

eslickj replied:
I guess my feeling (and I was too lazy to test this) is that named expressions have a value and it can be set.  It is deprecated and produces a warning message, but it still works.  I'll a little worried that if we don't explicitly limit this to variables or variables and parameters if that makes sense, someone could run into random weirdness at some point.

eslickj replied:
I should say I do use expressions in ports and probably the reason I'm not aware of this causing trouble is that the arrangement I have is probably always from expression to variable.

andrewlee94 replied:
@eslickj If you are using `Expressions` in `Ports`, that is the one time to use the `VarLikeExpression` component from IDAES as it was built specifically to address these issues. E.g. it overloads the `set_value` method to raise an `Exception` to prevent this from occurring. Also, `Expression` (or `VarLikeExpression` should (probably?) only be used in outlets, and not inlets because of this.

eslickj replied:
@andrewlee94 I do always use them in outlets, and I'm sure that's why I don't seem to have trouble with this.  Most properties aside from enthalpy, pressure, flow and composition in the Helmholtz EoS module are expressions.  I don't know what a person may put in a port or the direction they may think of their arcs going.  I guess a person could do something reasonable that we may think is weird.  That may argue for using VarLikeExpression for all expressions at least.  Maybe guarding against overwriting expressions here.

bknueven replied:
If we abandon the duck-typing with this change (as it is) the exception below will only be raised if the port has some non-Pyomo like object on it or the object incorrectly reports itself as a variable type. If the object implements `is_variable_type` and it returns `False`, it'll simply be ignored.

So if we expect the destinations to be variables I'd move the check `is_variable_type` up raise an exception if it's `False`, to avoid a perhaps unexpected silent failure.

eslickj commented:
Ok, I updated this to raise a TypeError if propagate_state tries to set anything that is not a variable.  It also raises a KeyError i guess if the variable you are setting is not indexed by a subset of the variable it's coming from.  I added a bunch of tests to to test pretty much all the combinations.
